realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -2474,3 +2486,480 @@ void TemperatureDiagnostics::diagnostics(diagnostic_updater::DiagnosticStatusWra
+            frameset.apply_filter(*named_filter._filter);
+            ROS_INFO("[NOMAGIC] Applied filter %s", named_filter._name.c_str());
+        }
+        frames_processed += 1;
+    }
+    ROS_INFO("[NOMAGIC] Filtered %d most recent frames", frames_processed);
+    ROS_INFO("[NOMAGIC] Frameset after filtering: %s", nomagicFramesetDescriptionString(frameset).c_str());
+    return frameset;
+
+}
+
+rs2::frame BaseRealSenseNode::nomagicGetDepthAlignedTo(stream_index_pair stream, rs2::frameset frameset)
+{
+    // This code is a modified and refactored chunk of publishAlignedDepthToOthers
+    ROS_INFO("[NOMAGIC] Computing depth aligned to %s%d from frameset %s",
+              rs2_stream_to_string(stream.first), stream.second, nomagicFramesetDescriptionString(frameset).c_str());
+
+    // Skip depth (makes no sense to align depth to depth)
+    // Skip 'sibling' streams like infra2, because ros-realsense does so for an unknown reason.
+    if (RS2_STREAM_DEPTH == stream.first || stream.second > 1) {
+        ROS_WARN("[NOMAGIC] Attempted an unexpected align to depth operation. Check source code for details");
+        return nomagicFramesetToFrame(stream, frameset);
+    }
+
+    // Make sure align filter is allocated.
+    std::shared_ptr<rs2::align> align;
+    try {
+        align = _align.at(stream.first);
+    }
+    catch(const std::out_of_range& e) {
+        ROS_DEBUG("[NOMAGIC] Allocated align filter for: %s_%d", rs2_stream_to_string(stream.first), stream.second);
+        align = (_align[stream.first] = std::make_shared<rs2::align>(stream.first));
+    }
+
+    // Apply alignment filter
+    rs2::frame aligned_depth = frameset.apply_filter(*align).as<rs2::frameset>().get_depth_frame();
+
+    // Apply colorizer filter if present
+    for (const auto & _filter : _filters) {
+        if (_filter._name == "colorizer") {
+            ROS_INFO("[NOMAGIC] Applying colorizer filter");
+            return _filter._filter->process(aligned_depth);
+        }
+    }
+
+    return aligned_depth;
+}
+
+sensor_msgs::ImagePtr BaseRealSenseNode::nomagicFrameToMessage(stream_index_pair stream, rs2::frame frame)
+{
+    // This code is a modified and refactored chunk of publishFrame.
+    assert(frame.is<rs2::video_frame>());
+
+    auto vframe = frame.as<rs2::video_frame>();
+    auto width = vframe.get_width();
+    auto height = vframe.get_height();
+    auto bpp = vframe.get_bytes_per_pixel();
+    auto& buffer = _image[stream];
+
+    buffer.create(height, width, buffer.type()); // This is lazy and won't reallocate if not needed.
+    buffer.data = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(frame.get_data()));
+
+    if (frame.is<rs2::depth_frame>()) {
+        // Note: fix_depth_scale returns it's second argument
+        buffer = fix_depth_scale(buffer, _depth_scaled_image[stream]);
+    }
+
+    sensor_msgs::ImagePtr img;
+    img = cv_bridge::CvImage(std_msgs::Header(), _encoding.at(stream.first), buffer).toImageMsg();
```

Comment:
Did you check this produced good RGB vs BGR metadata in Image message header? Just asking, since the Realsense node itself uses RBG, while OpenCV defaults to BGR.

Reply:
I tested this using RViz - the images on topics looked the same as from the service.